src/realm/object-store/sync/sync_user.hpp

```diff
@@ -228,26 +222,28 @@ class SyncUser : public std::enable_shared_from_this<SyncUser>, public Subscriba
     // Optionally set a context factory. If so, must be set before any sessions are created.
     static void set_binding_context_factory(SyncUserContextFactory factory);
 
-    std::shared_ptr<SyncManager> sync_manager() const REQUIRES(!m_mutex);
+    // Get the app instance that this user belongs to.
+    // This may not lock() if this SyncUser has become detached.
+    std::weak_ptr<app::App> app() const REQUIRES(!m_mutex);
 
     /// Retrieves a general-purpose service client for the Realm Cloud service
     /// @param service_name The name of the cluster
     app::MongoClient mongo_client(const std::string& service_name) REQUIRES(!m_mutex);
 
     // ------------------------------------------------------------------------
-    // All of the following are called by `SyncManager` and are public only for
+    // All of the following are called by `RealmBackingStore` and are public only for
     // testing purposes. SDKs should not call these directly in non-test code
     // or expose them in the public API.
 
-    explicit SyncUser(Private, const std::string& refresh_token, const std::string& id,
-                      const std::string& access_token, const std::string& device_id, SyncManager* sync_manager);
-    explicit SyncUser(Private, const SyncUserMetadata& data, SyncManager* sync_manager);
+    // Don't use this directly; use the `BackingStore` APIs. Public for use with `make_shared`.
+    SyncUser(std::string_view refresh_token, std::string_view id, std::string_view access_token,
+             std::string_view device_id, std::shared_ptr<app::App> app);
+    SyncUser(const SyncUserMetadata& data, std::shared_ptr<app::App> app);
```

Comment:
I had it public so that when SDKs make their own backing stores, those can create a SyncUser. But I tightened this up again by making it private, and allowing only child classes of a BackingStore to access it.